File: 152-maximum-product-subarray/152-maximum-product-subarray.cpp
class Solution {
public:
        int recur(vector<int>& nums,int i,bool takeIt,int cntNeg, vector<vector<vector<int>>>&memo){
        if(i>=nums.size()){
           return takeIt ? 1 : -1e6;
        }
        if(memo[i][takeIt][cntNeg%2]!=-1){
          //  cout<<nums[i]<<" , "<<memo[i][takeIt][cntNeg%2]<<"\n";
            return memo[i][takeIt][cntNeg%2];
        }
            
        if(takeIt){
          //  cout<<i<<" , "<<cntNeg<<"\n";
            if((cntNeg&1)==0){
                if(nums[i]<=0){
                 return memo[i][takeIt][cntNeg%2] = max(1,nums[i]*recur(nums,i+1,1,cntNeg+1,memo));
                }else{
                return memo[i][takeIt][cntNeg%2] = nums[i]*recur(nums,i+1,1,cntNeg,memo);
                }
            }else{
                if(nums[i]<=0)
                return memo[i][takeIt][cntNeg%2] = nums[i]*recur(nums,i+1,1,cntNeg+1,memo);
                return memo[i][takeIt][cntNeg%2] = nums[i]*recur(nums,i+1,1,cntNeg,memo);  
            }
        }
            
        int take = nums[i]*recur(nums,i+1,1,nums[i]<0?1:0,memo) ;
        int notTake = recur(nums,i+1,0,0,memo); 
            
        memo[i][takeIt][cntNeg%2]=max(take,notTake);
            
        return max(take,notTake);
    }
    int maxProduct(vector<int>& nums) {
    vector<vector<vector<int>>>memo(nums.size(),vector<vector<int>>(2,vector<int>(2,-1)));
          for (int i = 0; i <nums.size() ; ++i) {
        for (int j = 0; j <2 ; ++j) {
            for (int k = 0; k <2 ; ++k) {
                memo[i][j][k]=-1;
            }
        }
    }
        return recur(nums,0,0,0,memo);
    }
};

File: 152-maximum-product-subarray/152-maximum-product-subarray.java
class Solution {
    HashMap<String, Integer> dp;
    public int maxProduct(int[] nums) {
        int n = nums.length;
        dp = new HashMap();
        if(n == 0) return 0;
        return util(nums, n, 0, 1);
    }
    int util(int[] nums, int n, int index, int t){
        if(index == n) return t;
        String key = ""+index+t;
        if(dp.containsKey(key)) return dp.get(key);       
        
		// choice 1
        int x = util(nums, n, index+1, t*nums[index]);
		// choice 2
        int y = util(nums, n, index+1, nums[index]);
        if(index == 0){
            t = nums[0];
        }
		// keep the t (previous cal. value) and compare -> choice 3
        int ans = Math.max(t, Math.max(x, y));
        dp.put(key, ans);
        return ans;
    }
}